providers/providers.go

```diff
@@ -0,0 +1,196 @@
+package providers
+
+import (
+	"context"
+	"time"
+
+	flags "gx/ipfs/QmRMGdC6HKdLsPDABL9aXPDidrpmEHzJqFWSvshkbn9Hj8/go-ipfs-flags"
+	process "gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess"
+	procctx "gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess/context"
+	logging "gx/ipfs/QmTG23dvpBCBjqQwyDxV8CQT6jmS4PSftNr1VqHhE3MLy7/go-log"
+	routing "gx/ipfs/QmUHRKTeaoASDvDj7cTAXsmjAY7KQ13ErtzkQHZQq6uFUz/go-libp2p-routing"
+	peer "gx/ipfs/QmcJukH2sAFjY3HdBKq35WDzWoL3UUu2gt9wdfqZTUyM74/go-libp2p-peer"
+	cid "gx/ipfs/QmcZfnkapfECQGcLZaf9B79NRg7cRa9EnZh4LSbkCzwNvY/go-cid"
+	pstore "gx/ipfs/QmdeiKhUy1TVGBaKxt7y1QmBDLBdisSrLJ1x58Eoj4PXUh/go-libp2p-peerstore"
+	ipld "gx/ipfs/Qme5bWv7wtjUNGsK2BNGVUFPKiuxWrsqrtvYwCLRw8YFES/go-ipld-format"
+	host "gx/ipfs/QmfZTdmunzKzAGJrSvXXQbQ5kLLUiEMX5vdwux7iXkdk7D/go-libp2p-host"
+)
+
+const (
+	provideTimeout = time.Second * 15
+
+	// MaxProvidersPerRequest specifies the maximum number of providers desired
+	// from the network. This value is specified because the network streams
+	// results.
+	// TODO: if a 'non-nice' strategy is implemented, consider increasing this value
+	MaxProvidersPerRequest = 3
+	providerRequestTimeout = time.Second * 10
+
+	sizeBatchRequestChan = 32
+)
+
+var (
+	provideKeysBufferSize = 2048
+
+	// HasBlockBufferSize is the maximum numbers of CIDs that will get buffered
+	// for providing
+	HasBlockBufferSize = 256
+
+	provideWorkerMax = 512
+)
+
+var log = logging.Logger("providers")
+
+type blockRequest struct {
+	Cid *cid.Cid
+	Ctx context.Context
+}
+
+// Interface defines providers interface to libp2p routing system
+type Interface interface {
+	Provide(k *cid.Cid) error
+	ProvideRecursive(ctx context.Context, n ipld.Node, serv ipld.NodeGetter) error
+
+	FindProviders(ctx context.Context, k *cid.Cid) error
```

Comment:
It also does dht querying. I guess what you mean that instead of this method bitswap should use `FindProvidersAsync` and only send wantlists to those peers?

For now, I'd rename the method, and remove it as a part of other PR